File: Python/Word Search.py
class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
	def exist(self, board, word):
		self.totalRow, self.totalCol = len(board), len(board[0])
		for i in range(0,self.totalRow):
			for j in range(0,totalCol):
				if board[i][j] == word[0]:
					if self.dfs(board,i,j,word[1:]): return True
		return False
	def dfs(self,board,r,c,word):
		if len(word) == 0:return True
		if r>0 and board[r-1][c] == word[0]:
			ch, board[r][c] = board[r][c], "#"
			if self.dfs(board,r-1,c,word[1:]):return True
			board[r][c] = ch
		if r < self.totalRow-1 and board[r+1][c] == word[0]:
			ch, board[r][c] = board[r][c],"#"
			if self.dfs(board,r+1,c,word[1:]):return True
			board[r][c]=ch
		if c>0 and board[r][c-1] == word[0]:
			ch,board[r][c]=board[r][c],"#"
			if self.dfs(board,r,c-1,word[1:]):return True
			board[r][c] = ch
		if c < self.totalCol-1 and board[r][c+1] == word[0]:
			ch,board[r][c]=board[r][c],"#"
			if self.dfs(board,r,c+1,word[1:]):return True
			board[r][c] = ch
		return False

File: Java/Word Search.java
class Word Search{
	public boolean exist(char[][] board, String word) {
		if(board == null || board.length == 0 || word == null)
			return false;
		int rowLen = board.length;
		int colLen = board[0].length;
		for(int i=0;i<rowLen;i++){
			for(int j=0;j<colLen;j++){
				if(word.charAt(0) == board[i][j]){
					if(dfs(board,i,j,word.substring(1)))
						return true;
				}
			}
		}
		return false;
	}
	private boolean dfs(char[][] board, int row, int col, String s){
		if(s.length() == 0)
			return true;
		if(row > 0 && board[row-1][col] == s.charAt(0)){
			char c = board[row][col];
			board[row][col] = '$';
			if(dfs(board,row-1,col,s.substring(1)))
				return true;
			board[row][col] = c;
		}
		if(row<board.length-1 && board[row+1][col] == s.charAt(0)){
			char c = board[row][col];
			board[row][col] = '$';
			if(dfs(board,row+1,col,s.substring(1)))
				return true;
			board[row][col] = c;
		}
		if(col > 0 && board[row][col-1] == s.charAt(0)){
			char c = board[row][col];
			board[row][col] = '$';
			if(dfs(board,row,col-1,s.substring(1)))
				return true;
			board[row][col] = c;
		}
		if(col < board[0].length-1 && board[row][col+1] == s.charAt(0)){
			char c = board[row][col];
			board[row][col] = '$';
			if(dfs(board,row,col+1,s.substring(1)))
				return true;
			board[row][col] = c;
		}
		return false;
	}
}